c7n/resources/s3.py

```diff
@@ -3475,3 +3475,122 @@ def process_bucket(self, b):
+                filters:
+                  - type: bucket-replication
+                    state: True
+    """
+    schema = type_schema('bucket-replication',
+                         state={'type': 'boolean'},
+                         destination={'type': 'string', 'enum': ['cross-region', 'same-region']})
+
+    permissions = ("s3:GetReplicationConfiguration",)
+    annotation_key = 'c7n:bucket-replication'
+
+    def process(self, buckets, event=None):
+        results = []
+        with self.executor_factory(max_workers=2) as w:
+            futures = {w.submit(self.process_bucket, b): b for b in buckets}
+            for future in as_completed(futures):
+                b = futures[future]
+                if future.exception():
+                    self.log.error("Message: %s Bucket: %s", future.exception(),
+                                   b['Name'])
+                    continue
+                if future.result():
+                    results.append(b)
+        return results
+
+    def process_bucket(self, b):
+        client = bucket_client(local_session(self.manager.session_factory), b)
+        rules = []
+        if self.annotation_key not in b:
+            try:
+                bucket_replication = client.get_bucket_replication(Bucket=b['Name'])
+            except ClientError as e:
+                if e.response['Error']['Code'] != 'ReplicationConfigurationNotFoundError':
+                    raise
+                bucket_replication = {}
+            b[self.annotation_key] = bucket_replication
+        else:
+            bucket_replication = b[self.annotation_key]
+
+        rules = bucket_replication.get('ReplicationConfiguration', {}).get('Rules', [])
+        destination = self.data.get('destination')
+        # default `state` to True as previous impl assumed state == True
+        # to preserve backwards compatibility
+        if self.data.get('state', True):
+            if not destination:
+                if rules:
+                    return True
+                else:
+                    return False
+            else:
+                for replication in rules:
+                    if self.filter_bucket(b, replication, destination, client):
+                        return True
+                return False
+        else:
+            if not destination:
+                if rules:
+                    return False
+                else:
+                    return True
+            else:
+                for replication in rules:
+                    if not self.filter_bucket(b, replication, destination, client):
+                        return True
+                return False
+
+    def filter_bucket(self, b, replication, destination, client):
+        destination = self.data.get('destination')
+        destination_bucket = replication.get('Destination').get('Bucket').split(':')[5]
+        destination_region = client.get_bucket_location(
+            Bucket=destination_bucket).get('LocationConstraint')
```

Comment:
During today's community meeting, we had a question about how to find the endpoint URL that `inspect_bucket_region()` needs. You can get that from the client, for example:

```suggestion
        destination_region = inspect_bucket_region(destination_bucket, client.meta.endpoint_url)
```

This also avoids needing to remap `None` --> `us-east-1` and `EU` --> `eu-west-1` as we get the proper region names back in the first place.

Reply:
@ajkerrigan, thank you for the great suggestion. PR is updated.